generator/internal/sidekick/sidekick.go

```diff
@@ -17,57 +17,88 @@ package sidekick
+).
+	AddFlagString(&flagProjectRoot, "project-root", "", "the root of the output project").
+	AddFlagString(&format, "specification-format", "", "the specification format. Protobuf or OpenAPI v3.").
+	AddFlagString(&source, "specification-source", "", "the path to the input data").
+	AddFlagString(&serviceConfig, "service-config", "", "path to service config").
+	AddFlagString(&output, "output", "", "the path within project-root to put generated files").
+	AddFlagString(&flagLanguage, "language", "", "the generated language").
+	AddFlagBool(&dryrun, "dry-run", false, "do a dry-run: load the configuration, but do not perform any changes.").
+	AddFlagFunc("source-option", "source options", func(opt string) error {
+		components := strings.SplitN(opt, "=", 2)
+		if len(components) != 2 {
+			return fmt.Errorf("invalid source option, must be in key=value format (%s)", opt)
+		}
+		sourceOpts[components[0]] = components[1]
+		return nil
+	}).
+	AddFlagFunc("codec-option", "codec options", func(opt string) error {
+		components := strings.SplitN(opt, "=", 2)
+		if len(components) != 2 {
+			return fmt.Errorf("invalid codec option, must be in key=value format (%s)", opt)
+		}
+		codecOpts[components[0]] = components[1]
+		return nil
+	})
 
-func runSidekick(cmdLine *CommandLine) error {
-	if cmdLine.ProjectRoot != "" {
-		cwd, err := os.Getwd()
-		if err != nil {
-			return err
+// Run is the entry point for the sidekick logic. It expects args to be the command line arguments, minus the program name.
+func Run(args []string) error {
+	if len(args) < 1 {
+		_ = CmdSidekick.PrintUsage()
+		return fmt.Errorf("no command given")
+	}
+	if args[0] == "help" {
+		cmd, found, unusedArgs := CmdSidekick.Lookup(args[1:])
+		if !found {
+			return NotFoundError(cmd, args[1:], unusedArgs, fmt.Sprintf("Could not find help documentation for 'sidekick help %s'", strings.Join(args[1:], " ")))
 		}
-		defer os.Chdir(cwd)
-		if err := os.Chdir(cmdLine.ProjectRoot); err != nil {
+		return cmd.PrintUsage()
+	} else {
+		cmd, found, cmdArgs := CmdSidekick.Lookup(args)
+		if !found {
+			return NotFoundError(cmd, args, cmdArgs, fmt.Sprintf("Could not find command 'sidekick %s'", strings.Join(args, " ")))
+		} else {
+			var err error
+			if cmdLine, err := cmd.ParseCmdLine(cmdArgs); err == nil {
+				return runCommand(cmd, cmdLine)
+			}
 			return err
 		}
 	}
-	// Load the top-level configuration file. If there are any errors loading
-	// the file just run with the defaults.
-	rootConfig, err := loadRootConfig(".sidekick.toml")
-	if err != nil {
-		return err
-	}
-	argsConfig := &Config{
-		General: GeneralConfig{
-			Language: cmdLine.Language,
-		},
-		Source: maps.Clone(cmdLine.Source),
-		Codec:  maps.Clone(cmdLine.Codec),
-	}
-	config, err := mergeConfigs(rootConfig, argsConfig)
-	if err != nil {
-		return err
+}
+
+func NotFoundError(bestMatch *Command, allArgs []string, unusedArgs []string, msg string) error {
```

Comment:
maybe `constructNotFoundError`?

Reply:
used newNotFoundError, just to stay consistent with the newCommand function.